### PR TITLE
General maintenance for cryo chamber

### LINKED ISSUE #80699: Getting ejected from a cryo tube puts your sprite below the tube
## Issue Summary
Getting ejected from a cryo tube puts your sprite below the tube instead of above it

It's difficult to tell if someone is "under" a cryo tube due to the huge sprite
seems like a layer issue?

## Reproduction:
Get healed in a cryo tube, get ejected, your sprite is below the huge cryo tube sprite

### PR DESCRIPTION
**Pull request by SyncIt21**
## About The Pull Request
The changes can be summarized into the following categories

1. **Qol**(Should not be a problem for feature freeze hopefully)
   -  Adds balloon alerts for screwdriver & wrench acts, when inserting a beaker and other stuff
   - Adds extra examines for the beaker inside the cryo, how to open maintenance panel & other stuff. Most of the examines were not enclosed in `span_notice`
   - Adds screen tips for crowbar act, when inserting beaker into cryo etc
   - Cryo can now be pried open when there is no power with a crowbar so you can free someone stuck inside in case of an emergency
   - Cryo now auto turns on immediately without a 2 second delay & without the need for auto eject

2. **Fixes**   
    - Cryo won't drop the beaker onto the floor when ejected, but will put it in the players hands as intended
    - Fixes #80699. The cryo was on `ABOVE_MOB_LAYER` which rendered on top of everything, Now its set to the default layer so mobs & other stuff now render on top of it
    - Loc checks were incorrect instead of doing `mob.loc == src`(which would fail if the mob was standing in the open cro) we now do `get_turf(mob) == get_turf(src)`(which works if the mob by anyway in on the turf of the machine)

3. **Code Improvements**
   - Get's rid of unused variables such as `sleep_factor`, `unconscious_factor` etc. Simply unused
   - Autodocs remaining variables and procs
   - Slightly touched on the UI shared by cryo, chem heater etc. Removes the `transfer_amounts` var that went unused in these cases
 
4. **Refactors**
    - Converted UI to typescript
    - Removes `check_nap_violations()` from `process()`. This proc checks if the occupant has an ID card & money in the account which makes no sense for cryo
    - Cryo will turn off if there is no occupant OR if there is no beaker
    - It no longer starts processing round start & only when the machine is turned on/off thus saving cpu
    - Removed old cryo icon states such `cell-open` , `cell-off` etc. These were used for the old cryo tube from years ago but are no longer needed now so the icon file size has been trimmed

## Changelog
:cl:
qol: adds more examines, screentips & balloon alerts for cryo actions, cryo can be pried open with a crowbar when there is no power to free someone trapped inside
qol: cryo auto turns on immediately without a 2 second delay & without needing auto eject mode on
fix: ejecting beaker from cryotube will put it in the players hand & not drop it on the floor
fix: mobs & other stuff now render on top of cryo tubes and not bottom of it
fix: cryo checks for if the mob is on the same turf as cryo is now fixed, i.e. you can no longer close the machine on yourself
code: removed unused vars from cryo, autodocs vars, removed unused `transfer_amounts` from some chem machinery ui
refactor: removed unused icon states for cryo tube, cryo no longer processes round start and it's UI is now typescript
/:cl:

### CODE REVIEW

**Review by Wallemations — APPROVED** [review]
Those cryo sprites ain't used for the cloning museum piece, right?
If not, then go ahead.

**Comment by SyncIt21** [comment]
> Those cryo sprites ain't used for the cloning museum piece, right? If not, then go ahead.

Did a full project text search for those icon state's. Not one result popped up so definitely unused 👍

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 1]
```diff
@@ -445,25 +419,39 @@
...
+		balloon_alert(user, "turn off!")
+		return
 
-	var/unsafe_wrenching = FALSE
-	var/filled_pipe = FALSE
-	var/datum/gas_mixture/environment_air = loc.return_air()
+	var/unsafe_release = FALSE
 	var/datum/gas_mixture/inside_air = internal_connector.gas_connector.airs[1]
+	if(inside_air.total_moles() > 2 * ONE_ATMOSPHERE)
+		to_chat(user, span_warning("As you begin prying \the [src] a gush of air blows in your face... maybe you should reconsider?"))
```
Comment (FlufflesTheDog) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:443]: Does it make sense to expel the internal gas just for prying someone out? Also there's nothing stopping this from running if the machine is already open or lacking any occupants.

**Review by FlufflesTheDog — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 2]
```diff
@@ -239,110 +235,108 @@
...
-	if(!occupant)
-		return
+/obj/machinery/cryo_cell/end_processing()
+	. = ..()
+	SSair.stop_processing_machine(src)
 
-	if(!on)
-		// Should turn on if set to auto
-		if(autoeject)
-			set_on(TRUE)
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:274]: Have to find a way to make sure players don't close themselves in the cryo tube and it later turns on automatically. Will look into it

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 3]
```diff
@@ -239,110 +235,108 @@
...
-			return
+/obj/machinery/cryo_cell/on_set_is_operational(old_value)
+	//Turned off
+	if(old_value)
+		set_on(FALSE)
+
+/obj/machinery/cryo_cell/process(seconds_per_tick)
+	if(!occupant || !beaker)
+		set_on(FALSE)
+		return PROCESS_KILL
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:285]: The feedback is already present in the examines & screentips. Maybe before turning on the machine in the UI i can display it there, will do that

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 4]
```diff
@@ -445,25 +419,39 @@
...
+	if(on)
+		balloon_alert(user, "turn off!")
+		return
 
-	var/unsafe_wrenching = FALSE
-	var/filled_pipe = FALSE
-	var/datum/gas_mixture/environment_air = loc.return_air()
+	var/unsafe_release = FALSE
 	var/datum/gas_mixture/inside_air = internal_connector.gas_connector.airs[1]
+	if(inside_air.total_moles() > 2 * ONE_ATMOSPHERE)
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:442]: This isn't my code and was already there. But you caught a bug cool, pretty easy to fix

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 5]
```diff
@@ -472,138 +460,142 @@
...
 		if(beaker)
-			to_chat(user, span_warning("A beaker is already loaded into [src]!"))
+			balloon_alert(user, "beaker present!")
 			return
 		if(!user.transferItemToLoc(I, src))
 			return
 		beaker = I
-		user.visible_message(span_notice("[user] places [I] in [src]."), \
-							span_notice("You place [I] in [src]."))
+		balloon_alert(user, "beaker inserted!")
```
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:500]: Yeah good point

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 1 above)
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:443]: Yeah i need to add more checks to this.

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 4 above)
[the remark this answers is not in the document]
Comment (FlufflesTheDog) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:442]: Atmos code is kinda confusing but the old code looked correct with `if(internal_pressure > 2 * ONE_ATMOSPHERE)`

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:274]: I have a question though. If we are just goanna auto on the machine when the cryo tube is closed then why even have a "turn on" button in the UI?

[the remark this answers is not in the document]
Comment (FlufflesTheDog) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:274]: Auto-enable is off when cryo is set to manual I believe, though at that point I'm not entirely sure if the manual mode has much reason to exist either. Manual mode still automatically disables cryo for any of the normal reasons, which is kind of awkward for being "manual"

[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:274]: > manual mode has much reason to exist either

Manual mode though not convenient is much more preferred for balance reasons. If removed then players could just build cryo tubes and heal themselves anywhere without somebody's else intervention.

Would prefer to keep it over auto just for that

[the remark this answers is not in the document]
Comment (SyncIt21) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:274]: Maybe let's prefer manual mode more for balance reasons? I feel like auto mode would make the UI less useful (as the power button was one of the reasons to open it). I'll put it in the changelogs.

**Comment by RikuTheKiller** [comment]
Ok and you've made sure this won't cause 6+ bugs like the last "general maintenance" PR you made?

**Comment by SyncIt21** [comment]
> Why would you intentionally worsen QoL by making you have to power it on after closing it?

See https://github.com/tgstation/tgstation/pull/80873#discussion_r1450683771

> Why turn off cryotubes when they have no beaker inside?

Because cryo doesn't do any healing without a beaker containing Cryoxadone. No point wasting power otherwise

> Also, cryotubes can be used to trap people without healing them if the beaker is removed

Good point. I'll make sure to turn off the cryo & eject the occupant if the beaker is removed mid healing. Cryo should not be used to grief people by turning it into a death trap. That's not it's intended purpose, If we don't have a beaker of cryoxodane to do healing then operations should hault,

**Comment by FlufflesTheDog** [comment]
> Cryo should not be used to grief people by turning it into a death trap. That's not it's intended purpose, If we don't have a beaker of cryoxodane to do healing then operations should hault,

This sorta goes against removing the auto-enable qol
We're inevitably going to have waves of people that got used to that functionality and proceed to leave people in disabled cryo tubes for extended periods of time, now.

Also I don't know about how I feel about enforcing things should only be used for their intended purposes. Cryo is fantastic for stabilizing body temperatures of someone that's been on fire for example, or even lizards that are too cold, if you temporarily turn up the thermos. And death-trap things are just as important so antagonists have options to get creative with the sandbox.

(Although, being stuck asleep but unable to die in cryo is one of the least fun things to experience, so perhaps it shouldn't be TOO easy to accidentally do)

**Comment by SyncIt21** [comment]
Fair enough. I'll add auto enable qol again and remove the beaker checks for antagonists.

I guess no harm in having a redundant manual mode for beginners

**Comment by RikuTheKiller** [comment]
> Fair enough. I'll add auto enable qol again and remove the beaker checks for antagonists.
> 
> I guess no harm in having a redundant manual mode for beginners

Thanks, and personally I disliked your reasoning of "but muh UI" for making QoL worse.
A couple extra clicks every round doesn't change much except just annoying the docs and confusing newbies.

**Comment by RikuTheKiller** [comment]
> Cryo cannot be turned on without a occupant.

Why tho? It's not much, but it can be used to blow up a cryotube if the gas is hot enough.
Looks like yet another arbitrary lock, the atmos functionality of cryotubes works just fine even without an occupant.

**Comment by RikuTheKiller** [comment]
Also you forgot to update the "Refactors" part of the PR body to reflect the QoL and beaker changes being reverted.

**Comment by SyncIt21** [comment]
> Also you forgot to update the "Refactors" part of the PR body to reflect the QoL and beaker changes being reverted.

Updated

> Why tho? It's not much, but it can be used to blow up a cryotube if the gas is hot enough.

Cryo now works without a occupant. The cryo will still shutdown if it does not have any gas inside it but if it does then you can operate it with/without an occupant

**Comment by RikuTheKiller** [comment]
@Ghommie I thought we learned from the previous general maintenance PR that these conversions bundled with a bunch of fixes should be TM'd first...

Personally I've had PRs that change a couple messages get TM'd so I see no reason something like this wouldn't be.

Now look what you've done #80983 (not as bad as I expect it to be tbh)

**Comment by RikuTheKiller** [comment]
@SyncIt21 could you undo the "can't close on yourself" part of the PR and return it to how it was where it just closes without turning on if you do it on yourself? (at least that's how I remember it working)

**Comment by SyncIt21** [comment]
> @SyncIt21 could you undo the "can't close on yourself" part of the PR and return it to how it was where it just closes without turning on if you do it on yourself? (at least that's how I remember it working)

No. I would be undoing the changelog in #77919 which clearly states

> qol: Cryotubes will now automatically turn on when a patient enters it if auto is on, but you can no longer close the cryotube on yourself.

This previous behaviour

> where it just closes without turning on if you do it on yourself

was actually a bug and was never intended. The author used the wrong `loc` checks so they didn't enforce their own behaviour of 

> but you can no longer close the cryotube on yourself

I corrected that and now it does what that author initially wanted to do.

Any future discussions move it to the above linked issue. I won't be responding here

**Pull request merged**